cinn/hlir/framework/graph_compiler.cc

```diff
@@ -967,6 +972,95 @@ void GraphCompiler::RemoveInvalidVariables(const std::vector<std::unique_ptr<Ins
   });
 }
 
+static void cinn_buffer_malloc_with_callback(void* args, int num_args) {
+  cinn_pod_value_t* pod_args = reinterpret_cast<cinn_pod_value_t*>(args);
+  for (int i = 0; i < num_args; ++i) {
+    cinn_buffer_t* buffer = static_cast<cinn_buffer_t*>(pod_args[i]);
+    CHECK(buffer->external_malloc) << "external_malloc is nullptr";
+    buffer->external_malloc->operator()(nullptr, buffer);
+  }
+}
+
+static void cinn_buffer_free_with_callback(void* args, int num_args) {
+  cinn_pod_value_t* pod_args = reinterpret_cast<cinn_pod_value_t*>(args);
+  for (int i = 0; i < num_args; ++i) {
+    cinn_buffer_t* buffer = static_cast<cinn_buffer_t*>(pod_args[i]);
+    CHECK(buffer->external_free) << "external_free is nullptr";
+    buffer->external_free->operator()(nullptr, buffer);
+  }
+}
+
+void GraphCompiler::AnalysisVariableUsedLife(const std::vector<std::unique_ptr<Instruction>>& instructions,
```

Comment:
A better English name ... :

AnalyzeVariableLifeTime

Reply:
take it